posts/2019/09/ruby-wrappers-for-the-xnd-project.md

```diff
@@ -0,0 +1,388 @@
+<!--
+.. title: Ruby wrappers for the XND project
+.. slug: ruby-wrappers-for-the-xnd-project
+.. date: 2019-09-15 00:32:00 UTC-05:00
+.. tags: 
+.. category: 
+.. link: 
+.. description: 
+.. type: text
+-->
+
+<!-- markdown-toc start - Don't edit this section. Run M-x markdown-toc-generate-toc again -->
+**Table of Contents**
+
+- [Introduction](#introduction)
+- [Ndtypes](#ndtypes)
+    - [Usage](#usage)
+        - [Basic initialization](#basic-initialization)
+        - [Concrete Vs. Abstract Types](#concrete-vs-abstract-types)
+        - [Typedefs](#typedefs)
+        - [Usage via The C API](#usage-via-the-c-api)
+    - [Implementation](#implementation)
+- [Xnd](#xnd)
+    - [Basic Usage](#basic-usage)
+        - [Data Type Support](#data-type-support)
+        - [Missing Values](#missing-values)
+        - [Usage via The C API](#usage-via-the-c-api)
+    - [Implementation](#implementation)
+- [Gumath](#gumath)
+    - [Usage](#usage)
+        - [Usage via The C API](#usage-via-the-c-api)
+    - [Implementation](#implementation)
+    - [Automatic Kernel Generation](#automatic-kernel-generation)
+
+<!-- markdown-toc end -->
+
+
+# Introduction
+
+Lack of stable and reliable scientific computing software has been a persistent problem
+for the Ruby community, making it hard for enthusiastic Ruby developers to use Ruby in
+everything from their web applications to their data analysis projects. One of the most important
+components of any successful scientific software stack is a well maintained and flexible
+array computation library that can act as a fast and simple way of storing in-memory data
+and interfacing it with various fast and battle-tested libraries like LAPACK and BLAS.
+
+Various projects have attempted to make such libraries in the past (and some are still thriving
+and maintained). Some of the notable ones are [numo](https://github.com/ruby-numo), [nmatrix](https://github.com/SciRuby/nmatrix), and more recently, [numruby](https://github.com/SciRuby/numruby).
+These projects attempt to provide a simple Ruby-like API for creating and manipulating arrays
+of various types. All of them are able to easily interface with libraries like ATLAS, FFTW
+and LAPACK.
+
+However, all of the above projects fall short in two major aspects:
+* Lack of extensibility to adapt to modern use cases (read Machine Learning).
+* Lack of a critical mass of developers to maintain a robust and fast array library.
+
+The first problem is mainly due to the fact that they do not support very robust type systems.
+The available data types are limited and are extend to more complex uses. Modern use cases like
+Machine Learning require much more robust data types, as has been demonstrated by the tensor
```

Comment:
"more robust data types" --> maybe " a more robust type system" or "more data types (such as `bfloat16`)"? I'm still a little confused by what you mean by this sentence

Reply:
Like being able to define array shapes of arbitrary dimensions on multiple devices with a nice interface.